tico/passes/remove_redundant_reshape.py

```diff
@@ -427,3 +428,113 @@ def call(self, exported_program: ExportedProgram) -> PassResult:
+          P = aten.permute (R1, dims=(1, 0, 2))
+            ... arbitrary shape-preserving ops ...
+          R2 = aten.reshape(*, shape=(8, 20, 64))     # same as before
+
+        Effect:
+            * The first reshape loses one unit-dim.
+            * The permute dims are shifted down by 1 and the leading 0 is removed.
+        """
+        logger = logging.getLogger(__name__)
+
+        gm = ep.graph_module
+        graph = gm.graph
+        modified = False
+        for reshape_front in graph.nodes:
+            if not is_target_node(reshape_front, ops.aten.reshape):
+                continue
+            if len(reshape_front.users) != 1:
+                continue
+            # The reshape must actually add a leading 1.
+            reshape_front_shape = extract_shape(reshape_front)
+            if len(reshape_front_shape) != 0 and reshape_front_shape[0] != 1:
+                continue
+
+            permute = next(iter(reshape_front.users))
+            if not is_target_node(permute, ops.aten.permute):
+                continue
+
+            # The permute must keep that leading dim in front.
+            permute_shape = extract_shape(permute)
+            if permute_shape[0] != 1:
+                continue
+
+            # Walk through shape-preserving chain
+            tail = permute
+            while True:
+                if len(tail.users) != 1:
+                    break
+                nxt = next(iter(tail.users))
+                if is_target_node(nxt, ops.aten.reshape):
+                    break
+                if is_target_node(nxt, self.SHAPE_PRESERVING_OPS) and extract_shape(
+                    nxt
+                ) == extract_shape(tail):
+                    tail = nxt
+                    continue
+                break
+            if not is_target_node(nxt, ops.aten.reshape):
+                continue
+            reshape_back = nxt
+            # resahpe_back must drop exactly the leading 1.
+            reshape_back_shape = extract_shape(reshape_back)
+            if reshape_back_shape != permute_shape[1:]:
+                continue
+
+            permute_args = PermuteArgs(*permute.args, **permute.kwargs)  # type: ignore[arg-type]
+            with graph.inserting_before(permute):
+                new_front_shape = list(reshape_front_shape[1:])
+                reshape_new = graph.call_function(
+                    torch.ops.aten.reshape.default,
+                    args=(reshape_front.args[0], new_front_shape),
+                )
+                # Remove the first element and subtract 1 from every remaining index
+                new_dims = [d - 1 for d in permute_args.dims[1:]]
+                permute_new = graph.call_function(
+                    torch.ops.aten.permute.default, args=(reshape_new, new_dims)
+                )
+            permute.replace_all_uses_with(permute_new, propagate_meta=True)
+            tail_after_opt = tail if tail is not permute else permute_new
```

Comment:
May I ask a question?

isn't `tails` always `permute` because of next line..? 
https://github.com/Samsung/TICO/blob/60d404ca72a310a263c072e652cfaa2367cc0837/tico/passes/remove_redundant_reshape.py#L490-L491